### ISSUE TITLE
tests which time out don't get their tearDown() methods called

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2560]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#2560|
|Type|defect|
|Created|2007-04-09 22:46:33Z|

I don't know what the right solution to this is, but it causes problems that if a test times out, then it doesn't get a chance to tearDown(), thus disrupting the operation of successive tests.

<details><summary>Searchable metadata</summary>

```
trac-id__2560 2560
type__defect defect
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__trial trial
keywords__ 
time__1176158793000000 1176158793000000
changetime__1176810935000000 1176810935000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

So, timeout is an egregious hack that is deemed to evil to be used in actual application code. I wash my hands of this.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_2"></a>|@zooko commented|
|-|-|

I rather agree about timeouts, both in general and in specific.

So, if trial did not enforce timeouts on unit tests, then this would be consistent with your position.  (Test code could, of course, do timeouts itself if that's what it considered to be required for "passing".)

However, currently I have two problems:

1.  Trial is enforcing a 120 second timeout on one of my unit tests, even though I didn't ask it to.

2.  When that test fails due to timeout, trial doesn't give it a chance to tearDown(), so then other tests fail in weird ways.

Possible solutions:

1.  Trial could not enforce timeouts by default.  The code under test is still doing something bad, but using trial on bad code is good.

2.  Failure due to timeout could be treated the same as other kinds of failure, so that a test (which has a good tearDown()) which fails in any way leaves the other tests running well.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

I strongly prefer solution number 2.  Any unit test which cannot be made to pass in under 120 seconds has some extremely severe problems.  I can see wanting to work around those problems in some weird environment, and so it should be possible to tweak the timeout for a particular test, but this should be the exception, not the rule.  If anything, a default of 120 seconds is too generous; almost all tests should really be sub-second; more on the order of 10ms if they can help it.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_4"></a>|@zooko commented|
|-|-|

Glyph:

It kind of sounds like you want trial to (by default) enforce Twisted unit testing policy.  There's value in that, and but on the other hand perhaps trial would be more useful for more diverse users if it enabled people to easily configure that sort of policy for their specific project.

As a matter of fact, I do have a few unit tests which, when run on a heavily loaded Celery 500 MHz in parallel with a bunch of other buildslave tasks, take more than 120 seconds and then complete successfully.

Again, as a more experienced trial user I don't mind too much overriding the defaults for my test cases, but for other trial users, it might be good to make this knob more explicit so that they can more easily learn to control it.

(And I also intend to investigate why those unit tests take so long.  I suspect that they are disk bound.)

Anyway, I'm all in favor of solution 2.

Regards,

Zooko

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Replying to [zooko](#note_4):
> Glyph:
>
> It kind of sounds like you want trial to (by default) enforce Twisted unit testing policy.  There's value in that, and but on the other hand perhaps trial would be more useful for more diverse users if it enabled people to easily configure that sort of policy for their specific project.

This is not so much trying to enforce Twisted's unit testing policy, but setting some scope boundaries on Trial.  Trial is not a page-testing, functional testing, acceptance testing or integration testing framework.  It can be applied to all of these problems, since these definitions are not hard and fast, but no individual _unit_ of software should take more than two minutes for a single code path to be tested.  I believe it's a good design decision given Trial's domain to have timeouts by default.  In a different testing domain a different solution might be appropriate.

Even gene sequencing software can be tested on extremely short sequences to verify behavior on extremely long sequences.  Ray-tracers might take forever to render a complex image at high resolution, but _unit_ tests should be casting individual rays against particular objects, not rendering entire scenes.

> As a matter of fact, I do have a few unit tests which, when run on a heavily loaded Celery 500 MHz in parallel with a bunch of other buildslave tasks, take more than 120 seconds and then complete successfully.

I don't think that Trial should go out of its way to be brittle and inflexible, either.  These tests should be _far_ from the norm, and so specifying a specific timeout in such an outlying situation should not be a problem.

I've probably written quite a few trial tests like this myself, especially in Axiom, where the time required to commit a transaction can be off by two orders of magnitude depending on I/O load.  I still think those tests are busted and should be written in a better way.  (A ramfs fixture, maybe?)

> Again, as a more experienced trial user I don't mind too much overriding the defaults for my test cases, but for other trial users, it might be good to make this knob more explicit so that they can more easily learn to control it.

I'm not sure what you mean by "more explicit".  The interface I had in mind was "test_foo.timeout = 5000" or somesuch.  How would you make it more explicit than that?  Do you mean putting a "set .timeout if you expect this test to take this long" in the timeout error message?

> (And I also intend to investigate why those unit tests take so long.  I suspect that they are disk bound.)

As I said with the Axiom tests.  Maybe you want to investigate a ramfs fixture together? ;-)

> Anyway, I'm all in favor of solution 2.

Cool.  I guess I'm also proposing an implicit "solution 3" to combine with it: timeouts should be tweakable at a test level, rather than disabled by default.  It might even be reasonable to allow explicitly setting timeout = None.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

Before the discussion go too far, I must say that it works for me on trunk:

```
from twisted.trial import unittest
from twisted.internet import defer

class Test1(unittest.TestCase):
    def setUp(self):
        print "setup!"

    def tearDown(self):
        print "teardown"

    def test1(self):
        d = defer.Deferred()
        return d
    test1.timeout = 2
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [therve](#note_6):
> Before the discussion go too far, I must say that it works for me on trunk:

Hooray for time machines :).

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_8"></a>|@zooko commented|
|-|-|

Heh.  Your example works for me also.  Later today my friend who reported this bug to me will come to work and we'll find out if we actually have a bug at all.  Thanks, everyone.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set status to `closed`|
|-|-|

Thanks :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Yeah, several similar examples work for me both on trunk and in 2.5.  I didn't think to check, just assumed it was broken :).  Was there another bug lurking which made you think this?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml commented|
|-|-|

You mean apart from the lack of documentation in Trial and the floating "Arbitrarily timing out arbitrary Deferreds is bad mmmkay?" meme?

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)<a name="note_12"></a>|@warner commented|
|-|-|

I'm the one who told Zooko that timeouts bypassed tearDown, based upon the log messages I saw and a quick reading of the code.. my apologies if I was mistaken. I didn't think to run the simple test that therve mentioned. That test works correctly for me, as well as one that more accurately reflects our code by returning a Deferred from tearDown.

When I get a chance, I'll lower the timeouts on our long-running test and look more closely at the failure. The cleanup that is being done involves shutting down half a dozen Foolscap tubs, and it's possible that Tub.stopService() is not successfully terminating every connection and retiring every outstanding message. I think I finally nailed down the quiescent ones, but if the Tub is shutdown in the middle of heavy traffic, maybe something's getting missed.

Glyph: I hear what you're saying about trial being a *unit* test framework, and how it may not be the right tool for larger-scale system tests. (and the test that took hundreds of seconds for us, on an admittedly slow and overloaded machine, was our largest-scale SystemTest, doing an end-to-end check, simulating six independent nodes and hitting the highest-level APIs). But.. it's just so gosh-darn convenient :). It's great to have a buildbot just run trial and get all your failures in one easy-to-parse place.

In my mind, the timeout that trial offers/enforces is specifically to catch code that manages to drop a Deferred on the floor somewhere and thus forgets to fire one: no more progress can be made. Another expression of this is an infinite loop that includes the reactor, say something that does callLater() over and over again. (regular infinite loops never give the timeout a chance to fire, of course). The timeout exists to prevent trial (and, say, the buildslave running it) from running forever, so that the user who is running trial gets a chance to see the problem in their tests. (this is important because trial waits until the end of the run before it emits a number of very very useful messages, including the fact that it was a timeout which terminated a test). This timeout is valuable when tests are failing. Once they are passing, and you've nailed down all the pseudo-infinite-loops in the code, the timeout "should never happen" and can be as long as necessary.

Also, just to be clear, I'm not of the opinion that trial's timeouts should ever be used to assert that a particular test runs "quickly". We've had code in the past which was supposed to be, say, O(n), and an implementation error resulted in it being more like O(n^3), and once we fixed it, there was a temptation to have a unit test assert that we didn't accidentally go back to the old complexity. In my mind, any such unit test should really be measuring CPU seconds (or, ideally, VM opcode counts or something) rather than wallclock time. But of course that's infeasible, so my inclination would be to just remove such a test altogether.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

That makes me think that we would need some tools to do functional testing and performance testing. That may be using trial as a base, maybe not. But I think a tools that intends to to functional testing would be much better than using trial currently.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
